api/src/main/java/io/druid/data/input/impl/prefetch/JsonIterator.java

```diff
@@ -0,0 +1,141 @@
+import com.fasterxml.jackson.core.JsonToken;
+import com.fasterxml.jackson.core.ObjectCodec;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.druid.java.util.common.IAE;
+import io.druid.java.util.common.guava.CloseQuietly;
+import io.druid.java.util.common.io.Closer;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+/**
+ * An iterator over an array of JSON objects. Uses {@link ObjectCodec} to deserialize regular Java objects.
+ * @param <T> the type of object returned by this iterator
+ */
+public class JsonIterator<T> implements Iterator<T>, Closeable
+{
+  private JsonParser jp;
+  private ObjectCodec objectCodec;
+  private final TypeReference typeRef;
+  private final InputStream inputStream;
+  private final Closeable resourceCloser;
+  private final ObjectMapper objectMapper;
+
+  /**
+   *
+   * @param typeRef the object type that the JSON object should be deserialized into
+   * @param inputStream stream containing an array of JSON objects
+   * @param resourceCloser a {@code Closeable} implementation to release resources that the object is holding
+   * @param objectMapper object mapper, used for deserialization
+   */
+  public JsonIterator(
+      TypeReference typeRef,
+      InputStream inputStream,
+      Closeable resourceCloser,
+      ObjectMapper objectMapper
+  )
+  {
+    this.typeRef = typeRef;
+    this.inputStream = inputStream;
+    this.resourceCloser = resourceCloser;
+    jp = null;
+    this.objectMapper = objectMapper;
+  }
+
+  /**
+   * Returns {@code true} if there are more objects to be read.
+   * @return {@code true} if there are more objects to be read, else return {@code false}
+   */
+  @Override
+  public boolean hasNext()
+  {
+    init();
+    if (jp.isClosed()) {
+      return false;
+    }
+    if (jp.getCurrentToken() == JsonToken.END_ARRAY) {
+      CloseQuietly.close(jp);
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Retrieves the next deserialized object from the stream of JSON objects.
+   * @return the next deserialized object from the stream of JSON ovbjects
+   */
+  @Override
+  public T next()
+  {
+    if (!hasNext()) {
+      throw new NoSuchElementException("No more objects to read!");
+    }
+    try {
+      final T retVal = objectCodec.readValue(jp, typeRef);
+      jp.nextToken();
+      return retVal;
+    }
+    catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private void init()
+  {
+    if (jp == null) {
+      try {
+        if (inputStream == null) {
+          throw new UnsupportedOperationException();
+        } else {
+          jp = objectMapper.getFactory().createParser(inputStream);
+        }
+        final JsonToken nextToken = jp.nextToken();
+        if (nextToken == JsonToken.START_OBJECT || nextToken != JsonToken.START_ARRAY) {
```

Comment:
Do we need to check both conditions? Looks that `nextToken != JsonToken.START_ARRAY` is enough.

Reply:
Fixed.